README
-- DEPENDENCIES --
audio-convert-mod requires the following software to be installed:
  * python >= 2.4
  * PyGTK >= 2.10 with Glade support
  * Mutagen (often called python-mutagen)

Once audio-convert-mod is running, you may select "File > Show Format Conversion
Capabilities" to see which additional tools are required to convert audio into
your preferred format(s).

-- BUILDING --
  $ ./configure --prefix=/usr
  $ make

-- INSTALLING --
  $ make install

-- USEAGE --
To start audio-convert-mod, simply run:
  $ audio-convert-mod

If you would like to specify files to convert, add each an argument:
  $ audio-convert-mod /path/to/File1.mp3 /path/to/File2.ogg [...]

-- HELP --
If you think you've found a bug in audio-convert-mod, please send run the
program with verbose output enabled:
  $ audio-convert-mod --verbose
  
Then report a bug with the verbose messages _attached_, not posted please.


ChangeLog
|| Version 3.46.1 changes ||
  * Sat Aug 15 2009 Stewart Adam <s.adam at diffingo.com>
    - Add 'Resample' (= -1) to defaults when creating a new config
    - Fix a bug where encoding WAV files failed if resampling was enabled
    - Adjust position of popup menu when running on Darwin
    - Raise main window using GDKWindow.show() instead of hide()/show()

|| Version 3.46.0a changes ||
  * Fri Aug 14 2009 Stewart Adam <s.adam at diffingo.com>
    - Append '.converted' to filenames _before_ the extension
    - Fix a problem where encodes failed to complete if "Skip conversion" was
      not selected in the advanced options
    - Ensure that files added to the conversion list are handled internally
      using absolute paths

|| Version 3.46.0 changes ||
  * Mon Aug 3 2009 Stewart Adam <s.adam at diffingo.com>
    - Support audio resampling in advanced options
    - Fix a problem where selecting "Skip file" in the options would result in
      no files being encoded
    - Honor output folder override when self.decodeonly is True

  * Wed Jul 29 2009 Stewart Adam <s.adam at diffingo.com>
    - Redo autotools config files so that glib gettext is no longer required
    - Fix encode/decode capability detection for testbench
    - Use mutagen for reading and writing audio file tags
    - Fix a regression that made users unable to browse for folders in the
      conversion settings
    - Implement a logger similar to that of fwbackups to log a session's
      activity so that bug reporting and the subsequent troubleshooting is
      easier.
    - Fix several capability detection bugs when adding files to be converted

|| Version 3.45.5 changes ||
    - Suppots drag-and-drop of directories
    - Add support for KDE4 service menus in Dolphin and Konqueror
    - Add tag requirements to the Capabilities window
    - Create WAV files to a temporary directory instead of same folder as original
    - Add new "Remove Diacritis From Fileames" option
    - When adding files, filter out unsupported audio files
    - Use pynotify in place of blinking when possible
    - Bugfixes:
      - Call correct path when using a non-default prefix
      - Fix invalid usage when decoding MP3 files via Lame
      - Fix encoding to AC3 when output file already exists
      - When user has skipped a file, don't complain to the user about -9 retval
      - Detect tags correctly when certain fields are missing
    - Use internal error & warning dialogs instead of rat.hig
    - Add config.py for a better configuration file implementation
    - Update bitrates for OGG, MPC and AAC
    - Support AAC tags via libmp4v2
    - Add a test suite / "make check"

|| Version 3.45.4 changes ||
    - Fix bug when updating the preferences file (rh#442502 pt2)
    - Fix bug when loading defaults

|| Version 3.45.3 changes ||
    - Fix a bug where all file operations would fail when the file to convert
      contained quotes
    - Semi-Win* support... Coming soon!
    - Bugfix: The decode and encode fields were reversed in Features
    - Bugfix: WMA, SHA are NOT decodable for now.
    - New Chinese (zh_CN) translation (Thanks Harris!)
    - Move to trash is gone, use custom folder selection instead
    - Don't default to Output Folder --> Override
    - Fix OGG tagging (rh#442502)


|| Version 3.45.2 changes ||
    - Bugfix release ( no config directory = crash :/ )

|| Version 3.45.1 changes ||
    - Preferences / Save defaults
    - Tray icon
    - New options:
      * Output folder
      * Move originals to trash
    - Add more error checking
    
    || Version 3.45.0 changes ||
    - Rewrite into python
    - Write functionality of '.converted' script into main program
    - Support AC3, Wavpack
    - Extended tag/metadata functionality
    - Updated conversion commands (fewer glitches, proper progress)

|| Version 3.44 changes ||
    - Merged with audio-convert-Mmod (Modified version by Michael Schönitzer):
      - Made it possible to convert complete directories
      - Ask for files if none were given as argument
      - Check for every file, if it is supported
      - Made it possible to convert to the same filetype
      - Fixed German translation
      - Created a graphical installation-script
    - Improved "Remove .converted from files" script
    - Created a user uninstall script
    - No more system install... It was problematic and pretty useless
    - Support for KDE!
    - Internationalization!
    - Many, many bugfixes!
    - Now checks every file against supported formats
    - Folder and no folder menus

|| Version 3.43 changes ||
    - Fixed size of quality windows
    - Fixed OGG quality selection
    - Saved as UTF-8 to allow special characters
    - Fixed cancel functions; Hitting 'Cancel' now actually quits!
    - Fixed the 28 / 40 KB output file issues when converting from wav files.
    - Fixed the double-progress bar and the progress bar percentage when dealing with wav files
    - Fixed percentage when only encoding / decoding was needed
    - Fixed issues where the progress bar pulsed back and forth between two numbers
    - Optimized encode/decode functions
    - Change the two instances of FWaudio-convert-mod in copyright notices to the newer name

|| Version 3.42 changes ||
    - Added a chk-root function that will make a user root if one is not. (Thanks Jim for the function!)
    - Fixed many, many bugs in the installation and uninstall scripts
    - Fixed the name in the GPL headings: It still said audio-convert-mod.

|| Version 3.41 changes ||
    - Added 'data' file support (mplayer en/decoding)
    - Combined the Decode and Encode progress dialogs into one, less popup-age
    - Also combined ALL files into one progress bar, so even less popup-age!
    - Corrected labels to proper English, French
    - Script now simply adds '.convered' to a filename when it exists rather than asking to replace or adding .FW-ACMnew

src/audio_convert_mod/i18n.py
# -*- coding: utf-8 -*-
#  Copyright (C) 2007, 2008, 2009 Stewart Adam
#  This file is part of audio-convert-mod.

#  audio-convert-mod is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.

#  audio-convert-mod is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with audio-convert-mod; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
"""
This file is a slighlty modified version of the yum i18n.py file. Setups up the
audio-convert-mod translation domain and makes _() available. Uses ugettext to
make sure translated strings are in Unicode.
"""

try: 

  import gettext
  t = gettext.translation('audio-convert-mod')
  _ =  t.ugettext
except:
  # Oops! Better return the string as it is so we don't break things
  def _(str):
    """ [Wrapper] Returns the same string """
    return str


src/audio_convert_mod/const.py
# -*- coding: utf-8 -*-
#    Copyright (C) 2007, 2008, 2009 Stewart Adam
#    This file is part of audio-convert-mod.

#    audio-convert-mod is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.

#    audio-convert-mod is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with audio-convert-mod; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

"""
All constants/shared functions in the program
"""
import os
import sys
import getpass
import ConfigParser

MSWINDOWS = sys.platform.startswith("win")
LINUX = sys.platform.startswith('linux')
DARWIN = sys.platform == 'darwin'
if MSWINDOWS:
  UID = 1
  ROOTDRIVE = os.path.splitdrive(sys.argv[0])[0]
else:
  UID = os.getuid()
USER = getpass.getuser()

INSTALL_DIR = os.path.dirname(os.path.abspath(os.path.dirname(__file__)))
USERHOME = os.path.expanduser('~%s' % USER)
if USERHOME == '~%s' % USER: # if os doesn't support that format
  USERHOME = os.path.expanduser('~')
PREFLOC = os.path.normpath('%s/.audio-convert-mod/prefs.conf' % USERHOME)
LOGLOC = os.path.normpath('%s/.audio-convert-mod/userlog.txt' % USERHOME)

def which(program):
  """ Emulates unix `which` command """
  for path in os.getenv('PATH').split(os.pathsep):
    programPath = os.path.join(path, program)
    if os.path.exists(programPath) and os.path.isfile(programPath):
      return programPath
    if MSWINDOWS and os.path.isfile(programPath+'.exe'):
      return programPath
  return False




src/audio_convert_mod/widgets.py
# -*- coding: utf-8 -*-
#    Copyright (C) 2007, 2008, 2009 Stewart Adam
#    This file is part of audio-convert-mod.

#    audio-convert-mod is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.

#    audio-convert-mod is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with audio-convert-mod; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

"""
Custom widgets.
"""
import gtk
import gobject

import audio_convert_mod
from audio_convert_mod.i18n import _
from audio_convert_mod.const import *
from audio_convert_mod import formats

class StatusBar(gtk.Statusbar):
    def __init__(self, widget):
        """ Initialize.
            widget: GTK widget to use as the statusbar
        """
        gtk.Statusbar.__init__(self)
        self.statusbar = widget

    def newmessage(self, message, seconds=3):
        """ New message in status bar.
            message: Message to display
            seconds: number of seconds to display the message
                     defaults to three.
        """
        try:
            self.message_timeout()
        # Just means there was no existing message.
        except AttributeError:
            pass
        self.statusbar.push(1, message)
        self.message_timer = gobject.timeout_add(seconds*1000, self.message_timeout)

    def message_timeout(self):
        """ Remove a message from the statusbar """
        self.statusbar.pop(1)
        gobject.source_remove(self.message_timer)
        return True

class ProgressBar(gtk.ProgressBar):
    def __init__(self, widget, ms=15):
        """ Initialize.
            widget: GTK widget to use as the progressbar
            ms: Pulse step is every this many miliseconds
        """
        gtk.ProgressBar.__init__(self)
        self.progressbar = widget
        self.ms = ms
        self.pulsing = False
        self.progressbar.set_pulse_step(0.01)

    def _pulse(self):
        """ Pulse the progressbar. """
        self.progressbar.pulse()
        # Has to return true to happen again
        return True

    def startPulse(self):
        """ Start auto-pulsing the progressbar """
        if self.pulsing:
            self.stopPulse()
        self.pulsing = True
        self.pulsetimer = gobject.timeout_add(self.ms, self._pulse)

    def stopPulse(self):
        """ Stop auto-pulsing the progressbar """
        gobject.source_remove(self.pulsetimer)
        self.progressbar.set_fraction(0)
        self.pulsing = False
        return True

    def setMs(self, ms):
        """ Set the ms value (see __init__) """ 
        self.ms = int(ms)
        if self.pulsing:
            self.stopPulse()
            self.startPulse()

class GenericDia:
  """ Wrapper for the generic dialog """
  def __init__(self, dialog, title, parent):
    """ Initialize """
    self.dialog = dialog
    self.dialog.set_title(title)
    self.dialog.set_transient_for(parent)
    
  def run(self):
    """ Run the dialog """
    self.dialog.show()
    try:
      return self.dialog.run()
    except:
      self.destroy()
      raise
      
  def destroy(self):
    """ Hide the dialog, don't kill it! """
    self.dialog.hide()
    
  def runAndDestroy(self):
    """ Executes run() and destroy() """
    response = self.run()
    self.destroy()
    return response

class PathBrowser(GenericDia, gtk.FileChooserDialog):
  """ Wrapper for generic path dialogs """
  def __init__(self, dialog, parent, ffilter=None):
    """ Initialize.
      widget: The widget to use
      ffilter: List for filter creation.
          > 1st value is filter pattern
          > 2nd value is the description
    """
    GenericDia.__init__(self, dialog, _('Choose a file or folder'), parent)
    if ffilter:
      self.ffilter = gtk.FileFilter()
      for pattern in ffilter[:-1]:
        self.ffilter.add_pattern(pattern)
      self.ffilter.set_name(ffilter[-1])
      self.dialog.add_filter(self.ffilter)
    else:
      self.ffilter = None

  def destroy(self):
    """ Destroy the dialog and the filter """
    GenericDia.destroy(self)
    if self.ffilter:
      self.dialog.remove_filter(self.ffilter)
  
  def set_current_folder(self, path):
    """ Wrapper: See GTK+ help. """
    return self.dialog.set_current_folder(path)
  
  def set_select_multiple(self, multipleBool):
    """ Wrapper: See GTK+ help. """
    return self.dialog.set_select_multiple(multipleBool)

  def set_do_overwrite_confirmation(self, overwriteBool):
    """ Wrapper: See GTK+ help. """
    return self.dialog.set_do_overwrite_confirmation(overwriteBool)
    
  def set_action(self, action):
    """ Wrapper: See GTK+ help. """
    return self.dialog.set_action(action)
  
  def set_title(self, title):
    """ Wrapper: See GTK+ help. """
    return self.dialog.set_title(title)
  
  def get_filename(self):
    """ Wrapper: See GTK+ help. """
    return self.dialog.get_filenames()
  
  def get_filenames(self):
    """ Wrapper: See GTK+ help. """
    return self.dialog.get_filenames()
  
  def set_filename(self, filename):
    """ Wrapper: See GTK+ help. """
    return self.dialog.set_filename(filename)
  
  def add_filter(self, ffilter):
    """ Wrapper: See GTK+ help. """
    return self.dialog.add_filter(ffilter)
  
  def remove_filter(self, ffilter):
    """ Wrapper: See GTK+ help. """
    return self.dialog.remove_filter(ffilter)
  
# *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*

class bugReport(GenericDia):
  """ Wrapper for a bug report dialog """
  def __init__(self, dialog, textview, parent, tracebackText):
    """ Uses `dialog' to show bug report `traceback' in `textview' on top
        of `parent' """
    GenericDia.__init__(self, dialog, _('Bug Report'), parent)
    textview.get_buffer().set_text(tracebackText)

def saveFilename(parent):
  """ Displays a filechooser (save) and returns the chosen filename """
  fileChooser = gtk.FileChooserDialog(title=_('Save As...'),
                                      parent=parent,
                                      action=gtk.FILE_CHOOSER_ACTION_SAVE,
                                      buttons=(gtk.STOCK_CANCEL,
                                                 gtk.RESPONSE_CANCEL,
                                               gtk.STOCK_SAVE,
                                                 gtk.RESPONSE_OK))
  fileChooser.set_do_overwrite_confirmation(True)
  if fileChooser.run() in [gtk.RESPONSE_CANCEL, gtk.RESPONSE_DELETE_EVENT]:
    filename = None
  else:
    filename = fileChooser.get_filename()
  fileChooser.destroy()
  return filename



src/audio_convert_mod/acmlogger.py
# -*- coding: utf-8 -*-
#  Copyright (C) 2005, 2006, 2007, 2008, 2009 Stewart Adam
#  This file is part of audio-convert-mod.

#  audio-convert-mod is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.

#  audio-convert-mod is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with audio-convert-mod; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
"""
The audio-convert-mod logger
Based on fwbackups's fwlogger.py file (2009-07-29)
"""
import datetime
import logging
import types

from audio_convert_mod.const import *
from audio_convert_mod.i18n import _

L_DEBUG = logging.DEBUG
L_INFO = logging.INFO
L_WARNING = logging.WARNING
L_ERROR = logging.ERROR
L_CRITICAL = logging.CRITICAL
LOGGERS = {'main': 'acm-main'}
LEVELS = {'debug': 10,
          'info': 20,
          'warning': 30,
          'error': 40,
          'critical': 50}

def getLogger():
  """Retrieve the audio-convert-mod logger"""
  logging.setLoggerClass(acmLogger)
  logger = logging.getLogger(LOGGERS['main'])
  # reset to prevent excessive logging from other applications
  logging.setLoggerClass(logging.Logger)
  return logger

def shutdown():
  """Shut down the logging system"""
  logging.shutdown()

class acmLogger(logging.Logger):
  """A subclass to logging.Logger"""
  def __init__(self, name, level=logging.DEBUG):
    """Setup the audio-convert-mod logger, text mode"""
    logging.Logger.__init__(self, name, level)
    self.__printToo = False
    self.__functions = []
    self.__newmessages = False
    try:
      # need a handler
      loghandler = logging.FileHandler(LOGLOC, 'a')
      # Create formatter & add formatter to handler
      logformatter = logging.Formatter("%(message)s")
      loghandler.setFormatter(logformatter)
      # add handler to logger
      self.addHandler(loghandler)
    except Exception, error:
      print _('Could not set up the logger!')
      raise

  def setPrintToo(self, printToo):
    """If printToo is True, print messages to stdout as we log them"""
    self.__printToo = printToo
    
  def getPrintToo(self):
    """Retrieves the printToo property"""
    return self.__printToo
  
  def unconnect(self, function):
    """Disconnects a function from logmsg. Returns true if disconnected, false
        if that function was not connected."""
    try:
      self.__functions.remove(function)
      return True
    except ValueError:
      return False

  def connect(self, function):
    """Connects a function to logmsg. `function' must be passed as an instance,
        not the function() call itself.

        Function will be given the severity and message as arguments:
        def callback(severity, message)"""
    self.__functions.append(function)

  def logmsg(self, severity, message):
    """Logs a message. Severity is one of 'debug', 'info', 'warning', 'error'
    or 'critical'."""
    date = datetime.datetime.now().strftime('%b %d %H:%M:%S')
    level = self.getEffectiveLevel()
    if level <= LEVELS[severity.lower()]:
      entry = '%s :: %s : %s' % (date, _(severity.upper()), message)
      # pull in & execute the appropriate function
      getattr(self, severity.lower())(entry)
      if self.__printToo:
        print entry
      for i in self.__functions:
        i(severity.lower(), entry)



src/audio_convert_mod/interface.py
# -*- coding: utf-8 -*-
#  Copyright (C) 2007, 2008, 2009 Stewart Adam
#  This file is part of audio-convert-mod.

#  audio-convert-mod is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.

#  audio-convert-mod is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with audio-convert-mod; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
"""
Common classes for interfacing with Glade files
"""

import gtk
import gtk.glade
gtk.glade.bindtextdomain("audio-convert-mod")
# gtk
# |-- glade
#     |-- XML
#         |-- UserInterface     //Wraps a glade XML file
# Controller                    //Creates a UserInterface in 'ui'
# |-- audio-convert-mod                  //Starts program

class UserInterface(gtk.glade.XML):
  """Base class for UIs loaded from glade."""
  def __init__(self, filename, rootWidget, domain):
    """
      Initialize a new instance.
      `filename' is the name of the .glade file containing the UI hierarchy.
      `rootname' is the name of the topmost widget to be loaded.
      `gladeDir' is the name of the directory, relative to the Python
      path, in which to search for `filename'
    """
    gtk.glade.XML.__init__(self, filename, root=None, domain=domain)
    self.filename = filename
    self.root = self.get_widget(rootWidget)

  # *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*

  def __getattr__(self, name):
    """Look up an as-yet undefined attribute, assuming it's a widget."""
    result = self.get_widget(name)
    if result is None:
      raise AttributeError("Can't find widget %s in %s.\n" %
                 (`name`, `self.filename`))

    # Cache the widget to speed up future lookups.  If multiple
    # widgets in a hierarchy have the same name, the lookup
    # behavior is non-deterministic just as for libglade.
    setattr(self, name, result)
    return result

class Controller:
  """ Base class for all controllers of glade-derived UIs. """
  def __init__(self, gladeFile, rootWidget):
    """ Initialize a new instance.
      `gladeFile' is the glade XML file. """
    # Create and ui object contains the widgets.
    self.ui = UserInterface(gladeFile, rootWidget, 'audio-convert-mod')
    self.ui.signal_autoconnect(self._getAllMethods())

  def _getAllMethods(self):
    """ Get a dictionary of all methods in self's class hierarchy. """
    result = {}

    # Find all callable instance/class attributes.  This will miss
    # attributes which are "interpreted" via __getattr__.  By
    # convention such attributes should be listed in
    # self.__methods__.
    allAttrNames = self.__dict__.keys() + self._getAllClassAttributes()
    for name in allAttrNames:
      value = getattr(self, name)
      if callable(value):
        result[name] = value
    return result

  # *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*

  def _getAllClassAttributes(self):
    """ Get a list of all attribute names in self's class hierarchy. """
    nameSet = {}
    for currClass in self._getAllClasses():
      nameSet.update(currClass.__dict__)
    result = nameSet.keys()
    return result

  # *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*

  def _getAllClasses(self):
    """ Get all classes in self's heritage. """
    result = [self.__class__]
    i = 0
    while i < len(result):
      currClass = result[i]
      result.extend(list(currClass.__bases__))
      i = i + 1
    return result



src/audio_convert_mod/config.py
# -*- coding: utf-8 -*-
#  Copyright (C) 2005, 2006, 2007, 2008, 2009 Stewart Adam
#  This file is part of audio-convert-mod.

#  audio-convert-mod is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.

#  audio-convert-mod is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with audio-convert-mod; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
"""
Configuration classes for audio-convert-mod
"""
import ConfigParser
import os
import audio_convert_mod
from audio_convert_mod.const import *
from audio_convert_mod.i18n import _

class ConfigError(Exception):
  """ Errors in the configuration file. """
  def __init__(self, value):
    self.value = value
  def __str__(self):
    return repr(self.value)

def initConfigDir():
  """ Setup the configuration directory """
  configDir = os.path.normpath('%s/.audio-convert-mod' % USERHOME)
  if not os.path.exists(configDir):
    try:
      os.mkdir(configDir, 0755)
    except OSError, error:
      raise ConfigError(_('Could not create configuration directory: %s' % error))
      sys.exit(1)
  elif not audio_convert_mod.CheckPerms(configDir):
    raise ConfigError(_('You do not have read and write permissions to `%s\'.') % configDir)
    sys.exit(1)

class ConfigFile(ConfigParser.ConfigParser):
  """ A more sane ConfigParser. It commits changes immediately,
    and also re-parses after each action so
    what-you-see-is-what's-in-the-file.
  """
  # Basic setup.
  def __init__(self, conffile, create=False):
    ConfigParser.ConfigParser.__init__(self)
    self.config = ConfigParser.ConfigParser()
    ConfigParser.ConfigParser.optionxform = self.optionxform
    self.conffile = conffile
    self.create = create
    # Finally...
    self.read()

  def optionxform(self, option):
     """ Case sensitive! """
     return str(option)

  def set_conffile(self, conffile, create=True):
    """ Use a new config file.
      conffile: the path to the new configuration file
      create: If it doesn't exist, create a new config?
    """
    self.conffile = conffile
    self.read()

  def get_conffile(self):
    """ Returns the config file path. """
    return self.conffile

  def read(self):
    """ Read & parse the config file."""
    if self.create == True and not os.path.exists(self.conffile):
      self.fh = open(self.conffile,'w')
      self.fh.write('')
      self.fh.close()
    elif self.create == False and not os.path.exists(self.conffile):
      raise ConfigError(_('Attempted to read non-existant file \'%s\'') % self.conffile)
    self.fh = open(self.conffile,'r')
    self.config.readfp(self.fh)
    self.fh.close()

  def write(self):
    """ Write the config file, then reparse. """
    try:
      self.fh = open(self.conffile,'w')
      self.config.write(self.fh)
      self.fh.close()
      self.read()
    except:
      raise ConfigError(_('Cannot write to file \'%s\'') % self.conffile)

  def set(self, section, prop, value):
    """ Set a value in a given section and save. """
    self.config.set(section, prop, value)
    self.write()
    return True

  def get(self, section, prop):
    """ Returns the value for a given option in a section """
    value = self.config.get(section, prop)
    return value

  def remove_option(self, sectname, optionname):
    """ Remove a property & it's value, then save. """
    self.config.remove_option(sectname, optionname)
    self.write()
    return True

  def add_section(self, sectname):
    """ Add a section and save. """
    self.config.add_section(sectname)
    self.write()
    return True

  def remove_section(self, sectname):
    """ Remove a section and save. """
    self.config.remove_section(sectname)
    self.write()
    return True

  def sections(self):
    """ Returns the sections in the file """
    return self.config.sections()

  def options(self, section):
    """ Returns the options in a section of the file """
    return self.config.options(section)

  def has_section(self, section):
    """ Has section x? """
    return self.config.has_section(section)

  def has_option(self, section, option):
    """ Has option x in section y? """
    return self.config.has_option(section, option)

class PreferencesConf(ConfigFile):
  """ Preferences configuration file """
  def __init__(self, create=True):
    """ Checks for previous versions, incomplete confs, etc """
    ConfigFile.__init__(self, PREFLOC, create)
    if self.sections() == []:
      self.initialize()
    for i in ['Preferences', 'Defaults']:
      if not self.has_section(i):
        raise ConfigError(_('Configuration file is missing section \'%(a)s\'!') % {'a': i})
    self.patchConfiguration()

  def initialize(self):
    """ Creates a basic set configuration file """
    # Preferences section
    self.add_section('Preferences')
    import tempfile
    for pair in [['ShowTrayIcon', 1], ['TrayIconNotifications', 1], ['PauseOnErrors', 1],
                 ['TemporaryDirectory', tempfile.gettempdir()]]:
      self.set('Preferences', pair[0], pair[1])
    
    # Defaults section
    self.add_section('Defaults')
    # FileExists=1 --> append
    # Metadata=1 --> Copy
    # successfulConversion=1 --> leave converted files inplace
    # outputFolder=Off --> save to home
    for pair in [['Format', 'WAV'], ['Quality', 0], ['Extension', 0],
                ['FileExists', 1], ['Metadata', 1], ['SuccessfulConversion', 1],
                ['SuccessfulConversionDest', USERHOME], ['OutputFolder', 'Off'],
                ['RemoveDiacritics', 0], ['Resample', -1]]:
      self.set('Defaults', pair[0], pair[1])
    self.set('Preferences', 'Version', audio_convert_mod.__version__)

  def patchConfiguration(self):
    """ Import old configurations. Only runs if not current version """
    if not self.has_option('Preferences', 'Version'):
      # version was added in 3.45.1 so we import everything
      self.set('Preferences', 'Version', '3.45.0')
    try:
      oldVersion = self.get('Preferences', 'Version')
    except: # old config from beta3
      oldVersion = self.get('Preferences', 'version')

    # only if it's a version mismatch should we import
    if oldVersion == audio_convert_mod.__version__:
      return True

    # This lets us do what needs to be done for version X and _above_
    fromHereUp = False

    if oldVersion != audio_convert_mod.__version__:
      # This lets us do what needs to be done for version X and _above_
      fromHereUp = False
      # --------------------------------------------------------
      # oldest first, newest last in for the order of IF clauses
      # --------------------------------------------------------
      if oldVersion == '3.45.0' or fromHereUp == True:
        fromHereUp = True
        if not self.has_section('Defaults'):
          self.add_section('Defaults')
        for pair in [['format', 'WAV'], ['quality', 0], ['extension', 0],
                ['fileexists', 1], ['metadata', 1], ['successfulconversion', 1],
                ['successfulconversiondest', USERHOME], ['outputfolder', 'Off']]:
          self.set('Defaults', pair[0], pair[1])
      
      if oldVersion == '3.45.1' or fromHereUp == True:
        fromHereUp = True
      if oldVersion == '3.45.2' or fromHereUp == True:
        fromHereUp = True
        self.set('Defaults', 'successfulconversiondest', USERHOME)
      if oldVersion == '3.45.3' or fromHereUp == True:
        fromHereUp = True
        if not self.has_option('Defaults', 'successfulconversiondest'): # fix oops
          self.set('Defaults', 'successfulconversiondest', USERHOME)
      if oldVersion == '3.45.3' or fromHereUp == True:
        fromHereUp = True
      if oldVersion == '3.45.4' or fromHereUp == True:
        fromHereUp = True
        # Case sensitivity
        for option in ['Version', 'ShowTrayIcon', 'BlinkTrayIcon', 'PauseOnErrors']:
          self.set('Preferences', option, self.get('Preferences', option.lower()))
          self.remove_option('Preferences', option.lower())
        for option in ['Format', 'Quality', 'Extension', 'FileExists', 'Metadata', 'SuccessfulConversion', 'SuccessfulConversionDest', 'OutputFolder']:
          self.set('Defaults', option, self.get('Defaults', option.lower()))
          self.remove_option('Defaults', option.lower())
        # rename BlinkTrayicon
        self.set('Preferences', 'TrayIconNotifications', self.get('Preferences', 'BlinkTrayIcon'))
        self.remove_option('Preferences', 'BlinkTrayIcon')
        # add "remove diacritics" option
        self.set('Defaults', 'RemoveDiacritics', 0)
        # add temp dir option
        import tempfile
        self.set('Defaults', 'TemporaryDirectory', tempfile.gettempdir())
      if oldVersion == '3.45.5' or fromHereUp == True:
        fromHereUp = True
        self.set('Defaults', 'Resample', -1)
      if oldVersion in ['3.46.0', '3.46.0a'] or fromHereUp == True:
        fromHereUp = True
        
    return self.set('Preferences', 'Version', audio_convert_mod.__version__)

